torch_xla/experimental/distributed_checkpoint.py

```diff
@@ -45,26 +53,108 @@
+
+    # Flattend state_dict tracking all other state_dict items
+    self.unsharded_state_dict: Dict[str, Any] = None
+
+    # Upon the first `resolve_data` call for a WriteItem associated with a
+    # sharded tensor, all local shards are moved to CPU via
+    # `XLAShardedTensor::local_shards` and are tracked in `_local_shards` until
+    # the shard's data is resolved.
+    self._local_shards: Dict[str, List[XLAShard]] = {}
+
   def set_up_planner(self, state_dict: STATE_DICT_TYPE,
                      is_coordinator: bool) -> None:
-    raise NotImplemented
+    self.is_coordinator = is_coordinator
+
+    # Flatten the state_dict to allow separating sharded XLA tensors from
+    # types that can be handled by the default planner, and ensure all sharded
+    # tensors are wrapped in XLAShardedTensor
+    state_dict, self.mappings = flatten_state_dict(state_dict)
+    state_dict = tree_map(xs.wrap_if_sharded, state_dict)
+
+    # Select only XLAShardedTensors which are not replicated, since the
+    # default planner can handle everything else.
+    self.sharded_state_dict = {
+        k: v for k, v in state_dict.items() if _is_sharded_tensor(v)
+    }
+    unsharded = dict(state_dict.items() - self.sharded_state_dict.items())
+    self.unsharded_state_dict = tree_map(_unwrap_xla_sharded_tensor, unsharded)
 
   def create_local_plan(self) -> SavePlan:
-    raise NotImplemented
+    # Create the save plan for unsharded data
+    plan = create_default_local_save_plan(self.unsharded_state_dict,
+                                          self.is_coordinator)
+    # Track the flattened mappings in the plan metadata
+    plan = dataclasses.replace(plan, planner_data=self.mappings)
+
+    # Extend the plan for sharded tensor data
+    xla_write_items = _create_xla_write_items(self.sharded_state_dict)
+    plan.items.extend(xla_write_items)
+    return plan
 
   def create_global_plan(
       self, all_plans: List[SavePlan]) -> Tuple[List[SavePlan], Metadata]:
-    raise NotImplemented
+    # Deduplicate write items across plans
+    all_plans = dedup_tensors(all_plans)
+
+    global_plan, metadata = create_default_global_save_plan(all_plans)
+
+    # Combine mappings from all plans
+    planner_data_dict = [p.planner_data for p in global_plan]
+    merged_mappings = dict(ChainMap(*planner_data_dict))
+    metadata = dataclasses.replace(metadata, planner_data=merged_mappings)
+
+    return global_plan, metadata
 
   def finish_plan(self, new_plan: SavePlan) -> SavePlan:
-    raise NotImplemented
+    return new_plan
 
   def resolve_data(self,
                    write_item: WriteItem) -> Union[torch.Tensor, io.BytesIO]:
-    raise NotImplemented
+    obj = self.lookup_object(write_item.index)
+    return self.transform_object(write_item, obj)
+
+  def lookup_object(self, index: MetadataIndex) -> Any:
+    if index.fqn in self.unsharded_state_dict:
+      return self.unsharded_state_dict[index.fqn]
+
+    if index.fqn not in self._local_shards:
```

Comment:
It seems to me that the logic of having `_local_shards` is not necessary. You are basically just using it to keep track of duplicated writes. Will that happen? For LoadPlanner, it is necessary because you only want to load the shards to XLAShardedTensor when all the shards are in the host memory and then you need to count it. I'm not sure why the SavePlanner needs this logic. Am I missing anything?

Reply:
The motivation is that each call to `XLAShardedTensor::local_shards` will move all of the shards from device to CPU, and since we only need one shard at a time it's more efficient to only transfer once. I'll expand the comment in `__init__` to explain the need here.

Reply:
That makes sense to me now. Thanks for the explanation!